OMCompiler/SimulationRuntime/c/simulation/solver/newton_diagnostics.c

```diff
@@ -0,0 +1,1280 @@
+
+      r_x0 = fabs(f[i] + fz_dz);
+      if (r_x0 > maxRes)
+         maxRes = r_x0;
+   }
+
+   return maxRes;
+}
+
+// --------------------------------------------------------------------------------------------------------------------------------
+
+double*** getHessian( DATA* data, threadData_t *threadData, unsigned sysNumber, unsigned m)
+{
+   NONLINEAR_SYSTEM_DATA* systemData = &(data->simulationInfo->nonlinearSystemData[sysNumber]);
+   ANALYTIC_JACOBIAN* jac = &(data->simulationInfo->analyticJacobians[systemData->jacobianIndex]);
+
+   unsigned i, j, k;
+   const modelica_real eps = 1.e-7;
+   const modelica_real nominal_x = 1.e-4;
+   SIMULATION_DATA *sData = data->localData[0];
+
+   // Allocate memory for Hessian fxx (m * m * m doubles)
+   double*** fxx = (double***)malloc(m * sizeof(double**));
+   for( i = 0; i < m; i++)
+   {
+      fxx[i] = (double**)malloc(m * sizeof(double*));
+      for( j = 0; j < m; j++)
+         fxx[i][j] = (double*)malloc(m * sizeof(double));
+   }
+
+   // Allocate memory for Jacobians
+   double** fxPls = (double**)malloc(m * sizeof(double*));
+   double** fxMin = (double**)malloc(m * sizeof(double*));
+   for( i = 0; i < m; i++)
+   {
+      fxPls[i] = (double*)malloc(m * sizeof(double));
+      fxMin[i] = (double*)malloc(m * sizeof(double));
+   }
+
+   // ----------------------------------------------- Debug -------------------------------------------------
+   /*printf( "\n");
+   for( k = 0; k < m; k++)
+   {
+      unsigned id = var_id(k, data, systemData);
+      printf( "               k = %d: id = %d (%s)\n", k, id, data->modelData->realVarsData[id].info.name);
+   }*/
+   // -------------------------------------------- end of Debug ---------------------------------------------
+
+   for( k = 0; k < m; k++)
+   {
+      unsigned id = var_id(k, data, systemData);
+
+      double tmp_x = sData->realVars[id];
+      const modelica_real delta_x = eps * max( fabs(tmp_x), nominal_x);
+
+      sData->realVars[id] = tmp_x + delta_x;
+      for( j = 0; j < m; j++)
+      {
+         jac->seedVars[j] = 1.0;
+         ((systemData->analyticalJacobianColumn))(data, threadData, jac, NULL);
+         for( i = 0; i < m; i++)
+            fxPls[i][j] = jac->resultVars[i];
+         jac->seedVars[j] = 0.0;
+      }
+
+      sData->realVars[id] = tmp_x - delta_x;
+      for( j = 0; j < m; j++)
+      {
+         jac->seedVars[j] = 1.0;
+         ((systemData->analyticalJacobianColumn))(data, threadData, jac, NULL);
+         for( i = 0; i < m; i++)
+            fxMin[i][j] = jac->resultVars[i];
+         jac->seedVars[j] = 0.0;
+      }
+
+      sData->realVars[id] = tmp_x;
+
+      for( j = 0; j < m; j++)
+         for( i = 0; i < m; i++)
+         {
+            fxx[i][k][j] = (fxPls[i][j] - fxMin[i][j]) / (2 * delta_x);
+            if (isnan(fxx[i][k][j]))
+            {
+               printf("NaN detected: fxx[%d][%d][%d]: fxPls[%d][%d] = %f, fxMin[%d][%d] = %f, delta_x = %f\n",
```

Comment:
Don't use `printf` and similar in the C runtime. The print functions `infoStreamPrint` can interact with OMEdit and
only print if the specified debug flag is active.

You can use

```C
infoStreamPrint(LOG_STDOUT, 0, "Stuff");
```

to print nearly always or even better only when `-ls=LOG_NLS_NEWTON_DIAG` is provided:

```C
infoStreamPrint(LOG_NLS_NEWTON_DIAG, 0, "Stuff");
```